File: apps/etl-clojure/src/pulso/xml/transform.clj
(ns pulso.xml.transform
  (:import [java.time OffsetDateTime LocalDate]
           [java.time.format DateTimeFormatter]))

(def ^:private apple-health-datetime-fmt
  (DateTimeFormatter/ofPattern "yyyy-MM-dd HH:mm:ss Z"))

(defn parse-datetime
  "Parses Apple Health datetime string to OffsetDateTime."
  [s]
  (when (and s (not (empty? s)))
    (OffsetDateTime/parse s apple-health-datetime-fmt)))

(defn parse-date
  "Parses a date string (yyyy-MM-dd) to LocalDate."
  [s]
  (when (and s (not (empty? s)))
    (LocalDate/parse s)))

(defn parse-dbl
  "Safely parses a string to Double, returns nil on failure."
  [s]
  (when (and s (not (empty? s)))
    (try (Double/parseDouble s) (catch NumberFormatException _ nil))))

(defn- extract-metadata
  "Extracts MetadataEntry children from an element."
  [element]
  (->> (:content element)
       (filter #(= (:tag %) :MetadataEntry))
       (mapv (fn [entry]
               {:key   (get-in entry [:attrs :key])
                :value (get-in entry [:attrs :value])}))))

(defn export-date-element->map [element]
  {:export-date (parse-datetime (get-in element [:attrs :value]))})

(defn me-element->map [element locale]
  (let [attrs (:attrs element)]
    {:date-of-birth       (parse-date (:HKCharacteristicTypeIdentifierDateOfBirth attrs))
     :biological-sex      (:HKCharacteristicTypeIdentifierBiologicalSex attrs)
     :blood-type          (:HKCharacteristicTypeIdentifierBloodType attrs)
     :fitzpatrick-skin    (:HKCharacteristicTypeIdentifierFitzpatrickSkinType attrs)
     :cardio-fitness-meds (:HKCharacteristicTypeIdentifierCardioFitnessMedicationsUse attrs)
     :locale              locale}))

(defn record-element->map [element]
  (let [attrs (:attrs element)]
    {:type           (:type attrs)
     :source-name    (:sourceName attrs)
     :source-version (:sourceVersion attrs)
     :unit           (:unit attrs)
     :value          (:value attrs)
     :device         (:device attrs)
     :creation-date  (parse-datetime (:creationDate attrs))
     :start-date     (parse-datetime (:startDate attrs))
     :end-date       (parse-datetime (:endDate attrs))
     :metadata       (extract-metadata element)}))

(defn workout-element->map [element]
  (let [attrs (:attrs element)
        children (:content element)]
    {:activity-type            (:workoutActivityType attrs)
     :duration                 (parse-dbl (:duration attrs))
     :duration-unit            (:durationUnit attrs)
     :total-distance           (parse-dbl (:totalDistance attrs))
     :total-distance-unit      (:totalDistanceUnit attrs)
     :total-energy-burned      (parse-dbl (:totalEnergyBurned attrs))
     :total-energy-burned-unit (:totalEnergyBurnedUnit attrs)
     :source-name              (:sourceName attrs)
     :source-version           (:sourceVersion attrs)
     :device                   (:device attrs)
     :creation-date            (parse-datetime (:creationDate attrs))
     :start-date               (parse-datetime (:startDate attrs))
     :end-date                 (parse-datetime (:endDate attrs))
     :metadata                 (extract-metadata element)
     :events                   (->> children
                                    (filter #(= (:tag %) :WorkoutEvent))
                                    (mapv (fn [e]
                                            (let [a (:attrs e)]
                                              {:type          (:type a)
                                               :date          (parse-datetime (:date a))
                                               :duration      (parse-dbl (:duration a))
                                               :duration-unit (:durationUnit a)}))))
     :statistics               (->> children
                                    (filter #(= (:tag %) :WorkoutStatistics))
                                    (mapv (fn [e]
                                            (let [a (:attrs e)]
                                              {:type       (:type a)
                                               :start-date (parse-datetime (:startDate a))
                                               :end-date   (parse-datetime (:endDate a))
                                               :average    (parse-dbl (:average a))
                                               :minimum    (parse-dbl (:minimum a))
                                               :maximum    (parse-dbl (:maximum a))
                                               :sum        (parse-dbl (:sum a))
                                               :unit       (:unit a)}))))
     :routes                   (->> children
                                    (filter #(= (:tag %) :WorkoutRoute))
                                    (mapv (fn [e]
                                            (let [a (:attrs e)]
                                              {:source-name (:sourceName a)
                                               :start-date  (parse-datetime (:startDate a))
                                               :end-date    (parse-datetime (:endDate a))
                                               :file-path   (->> (:content e)
                                                                  (filter #(= (:tag %) :FileReference))
                                                                  first
                                                                  :attrs
                                                                  :path)}))))}))

(defn correlation-element->map [element]
  (let [attrs (:attrs element)]
    {:type           (:type attrs)
     :source-name    (:sourceName attrs)
     :source-version (:sourceVersion attrs)
     :device         (:device attrs)
     :creation-date  (parse-datetime (:creationDate attrs))
     :start-date     (parse-datetime (:startDate attrs))
     :end-date       (parse-datetime (:endDate attrs))
     :metadata       (extract-metadata element)
     :records        (->> (:content element)
                          (filter #(= (:tag %) :Record))
                          (mapv record-element->map))}))

(defn activity-summary-element->map [element]
  (let [attrs (:attrs element)]
    {:date-components           (parse-date (:dateComponents attrs))
     :active-energy-burned      (parse-dbl (:activeEnergyBurned attrs))
     :active-energy-burned-goal (parse-dbl (:activeEnergyBurnedGoal attrs))
     :active-energy-burned-unit (:activeEnergyBurnedUnit attrs)
     :apple-move-time           (parse-dbl (:appleMoveTime attrs))
     :apple-move-time-goal      (parse-dbl (:appleMoveTimeGoal attrs))
     :apple-exercise-time       (parse-dbl (:appleExerciseTime attrs))
     :apple-exercise-time-goal  (parse-dbl (:appleExerciseTimeGoal attrs))
     :apple-stand-hours         (parse-dbl (:appleStandHours attrs))
     :apple-stand-hours-goal    (parse-dbl (:appleStandHoursGoal attrs))}))

File: apps/etl-python/pulso/xml/transform.py
from datetime import datetime, date

DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S %z"


def parse_datetime(s):
    if not s:
        return None
    return datetime.strptime(s, DATETIME_FORMAT)


def parse_date(s):
    if not s:
        return None
    return date.fromisoformat(s)


def parse_dbl(s):
    if not s:
        return None
    try:
        return float(s)
    except ValueError:
        return None


def _extract_metadata(element):
    return [
        {"key": e.attrib.get("key"), "value": e.attrib.get("value")}
        for e in element.findall("MetadataEntry")
    ]


def export_date_element_to_map(element):
    return {"export_date": parse_datetime(element.attrib.get("value"))}


def me_element_to_map(element, locale):
    a = element.attrib
    return {
        "date_of_birth": parse_date(a.get("HKCharacteristicTypeIdentifierDateOfBirth")),
        "biological_sex": a.get("HKCharacteristicTypeIdentifierBiologicalSex"),
        "blood_type": a.get("HKCharacteristicTypeIdentifierBloodType"),
        "fitzpatrick_skin": a.get("HKCharacteristicTypeIdentifierFitzpatrickSkinType"),
        "cardio_fitness_meds": a.get("HKCharacteristicTypeIdentifierCardioFitnessMedicationsUse"),
        "locale": locale,
    }


def record_element_to_map(element):
    a = element.attrib
    return {
        "type": a.get("type"),
        "source_name": a.get("sourceName"),
        "source_version": a.get("sourceVersion"),
        "unit": a.get("unit"),
        "value": a.get("value"),
        "device": a.get("device"),
        "creation_date": parse_datetime(a.get("creationDate")),
        "start_date": parse_datetime(a.get("startDate")),
        "end_date": parse_datetime(a.get("endDate")),
        "metadata": _extract_metadata(element),
    }


def workout_element_to_map(element):
    a = element.attrib

    events = [
        {
            "type": e.attrib.get("type"),
            "date": parse_datetime(e.attrib.get("date")),
            "duration": parse_dbl(e.attrib.get("duration")),
            "duration_unit": e.attrib.get("durationUnit"),
        }
        for e in element.findall("WorkoutEvent")
    ]

    statistics = [
        {
            "type": s.attrib.get("type"),
            "start_date": parse_datetime(s.attrib.get("startDate")),
            "end_date": parse_datetime(s.attrib.get("endDate")),
            "average": parse_dbl(s.attrib.get("average")),
            "minimum": parse_dbl(s.attrib.get("minimum")),
            "maximum": parse_dbl(s.attrib.get("maximum")),
            "sum": parse_dbl(s.attrib.get("sum")),
            "unit": s.attrib.get("unit"),
        }
        for s in element.findall("WorkoutStatistics")
    ]

    routes = []
    for r in element.findall("WorkoutRoute"):
        file_ref = r.find("FileReference")
        routes.append({
            "source_name": r.attrib.get("sourceName"),
            "start_date": parse_datetime(r.attrib.get("startDate")),
            "end_date": parse_datetime(r.attrib.get("endDate")),
            "file_path": file_ref.attrib.get("path") if file_ref is not None else None,
        })

    return {
        "activity_type": a.get("workoutActivityType"),
        "duration": parse_dbl(a.get("duration")),
        "duration_unit": a.get("durationUnit"),
        "total_distance": parse_dbl(a.get("totalDistance")),
        "total_distance_unit": a.get("totalDistanceUnit"),
        "total_energy_burned": parse_dbl(a.get("totalEnergyBurned")),
        "total_energy_burned_unit": a.get("totalEnergyBurnedUnit"),
        "source_name": a.get("sourceName"),
        "source_version": a.get("sourceVersion"),
        "device": a.get("device"),
        "creation_date": parse_datetime(a.get("creationDate")),
        "start_date": parse_datetime(a.get("startDate")),
        "end_date": parse_datetime(a.get("endDate")),
        "metadata": _extract_metadata(element),
        "events": events,
        "statistics": statistics,
        "routes": routes,
    }


def correlation_element_to_map(element):
    a = element.attrib
    return {
        "type": a.get("type"),
        "source_name": a.get("sourceName"),
        "source_version": a.get("sourceVersion"),
        "device": a.get("device"),
        "creation_date": parse_datetime(a.get("creationDate")),
        "start_date": parse_datetime(a.get("startDate")),
        "end_date": parse_datetime(a.get("endDate")),
        "metadata": _extract_metadata(element),
        "records": [record_element_to_map(r) for r in element.findall("Record")],
    }


def activity_summary_element_to_map(element):
    a = element.attrib
    return {
        "date_components": parse_date(a.get("dateComponents")),
        "active_energy_burned": parse_dbl(a.get("activeEnergyBurned")),
        "active_energy_burned_goal": parse_dbl(a.get("activeEnergyBurnedGoal")),
        "active_energy_burned_unit": a.get("activeEnergyBurnedUnit"),
        "apple_move_time": parse_dbl(a.get("appleMoveTime")),
        "apple_move_time_goal": parse_dbl(a.get("appleMoveTimeGoal")),
        "apple_exercise_time": parse_dbl(a.get("appleExerciseTime")),
        "apple_exercise_time_goal": parse_dbl(a.get("appleExerciseTimeGoal")),
        "apple_stand_hours": parse_dbl(a.get("appleStandHours")),
        "apple_stand_hours_goal": parse_dbl(a.get("appleStandHoursGoal")),
    }
